lib/identity_cache/query_api.rb

```diff
@@ -154,6 +148,25 @@ def cache_fetch_includes
       end
     end
 
+    included do |base|
+      # Make sure there is at least once after_commit callback so that _run_commit_callbacks
+      # is called, which is overridden to do an early after_commit callback
+      base.after_commit {}
+    end
+
+    # Override the method that is used to call after_commit callbacks so that we can
+    # expire the caches before other after_commit callbacks. This way we can avoid stale
+    # cache reads that happen from the ordering of callbacks. For example, if an after_commit
+    # callback enqueues a background job, then we don't want it to be possible for the
+    # background job to run and load data from the cache before it is invalidated.
+    def _run_commit_callbacks
```

Comment:
I'm assuming this is not quite a public RoR API. I worry about weird interactions with the main implementation, I guess we are just calling `super`. Maybe I'm just being paranoid .

Reply:
Looks like the ActiveSupport::Callbacks documentation says to call `run_callbacks` and it looks like [`_run_#{name}_callbacks` is just a wrapper around run_callbacks](https://github.com/rails/rails/blob/v6.0.3.3/activesupport/lib/active_support/callbacks.rb#L824-L826).  So it looks like `run_callbacks` would be the more public and stable API to override.  However, it is also probably an implementation detail that Active Record callbacks are implemented using `ActiveSupport::Callbacks` given their use of wrapper methods rather suggesting that `set_callback` be used directly to define these callbacks.